### ISSUE TITLE
"set_target_properties" not evaluated correctly: cannot open file '*-NOTFOUND.obj'

### ISSUE THREAD

**Issue by guestieng** [microsoft/vscode-cmake-tools#1443]
### Brief Issue Summary

Importing an external static library via "add_library" and "set_target_properties" leads to an unexpected linking error:
[build] LINK : fatal error LNK1104: cannot open file '*-NOTFOUND.obj'

### Expected:

1) In CMakeLists.txt define a target for an external static library (not a cmake project) and pass it to target_link_libraries:

```
add_library(Thislib STATIC IMPORTED)
set_target_properties(Thislib PROPERTIES
    IMPORTED_LOCATION "${THISLIB_STATIC_LIB}"
    INTERFACE_INCLUDE_DIRECTORIES "${THISLIB_INCLUDE_DIR}"
)

target_link_libraries( ${PROJECT_NAME}
    Thislib 
)
```
where THISLIB_STATIC_LIB and THISLIB_INCLUDE_DIR are ultimately defined in CMakeCache.txt .

2) Use Visual Studio Generator to generate, build and install everything. This works all fine, if done, e.g., within Visual Studio or manually.

### Apparent Behavior:

If doing step 2) from above within Visual Studio Code and with vscode-cmake-tools, the following linking error appears:

`[build] LINK : fatal error LNK1104: cannot open file 'Thislib-NOTFOUND.obj'`

However, looking into the Visual Studio project files one finds that all paths are set correctly. Besides, the static library is of course existing at the right location.

### Platform and Versions

- **Operating System**: Windows 10..0.16299
- **CMake Version**: 3.17.1
- **VSCode Version**: 1.48.0
- **CMake Tools Extension Version**: 1.4.2
- **Compiler/Toolchain**: Visual C++ 2019

**Comment by guestieng** [comment]
This particular issue can be considered as not being a bug actually. 
In this sense one would probably call it resolved.
However, what led to this issue was, after all, a matter of usability (provided the user is not required to be very attentive all the time):

If one works with (multiple) multi-root workspaces one easily runs into conflicts with a "cmake.buildDirectory" definition
which seemingly has(!) to be defined in a ".vscode/settings.json" within a given source directory, in order to make a multi-root workspace be usable properly with cmake-tools.
Now, to my taste this is far from an optimal place where to have such a definition, since everything you want in a source directory are solely the original sources and no IDE specific definitions which (in there "core") should be derivable by cmake data elsewhere. 
Is there not a way to have cmake-tools using Cmake more directly?

**Comment by bobbrow** [comment]
It's my opinion that the `.vscode/settings.json` should generally not be committed to source control and it makes sense to include it in your .gitignore.  But we have default values for these settings and you shouldn't need to modify them to make the extension work.

The more relevant issue here seems to be that you have data in your CMakeCache that you want the extension to use instead of these settings.  Is that correct?  Can you share more info about your CMake project structure and which folders are open in your multi-root workspace?  Perhaps that can help us understand your scenario a bit better.

Thank you!

**Comment by guestieng** [comment]
That's true, `.vscode/settings.json` should go in the .gitignore file. Nevertheless it is an extra step; and, it may be a matter of taste indeed, but I like a source directory or, respectively, the `.vscode/settings.json` file to be as clean as possible and have relevant editor related definitions scattered as little as possible around the entire workspace.

Regarding the default settings: they cover a lot of use cases indeed, but not all; one example are varying path patterns in multi-root workspaces. Hence, in order to keep as much freedom and direct control over individual projects as possible within VSCode, the only way I found working is to generically (and explicitly!) define "cmake.buildDirectory" in `.vscode/settings.json` files.

So, to answer your question ("...  Is that correct?"), yes,
I wonder, whether it would not be better (as to covering a bigger set of workflows in a more general and efficient way) to exlpoit CMakeCache more, if you want. 

... By now, the multi-root workspaces are defined via the "folders" array in *.code-workspace files. These are located in corresponding build directories. But in order to have everything working together smoothly, the projects source paths need to be defined in that json "folders" array. Thus, to my understanding, `cmake.buildDirectory ` is of course needed to "close the circle" and provide information about the build directories. 
But unfortunately this whole procedure (and only this one worked for having the above features) seems like being the reverse of what I would intuitively expect on the basis of cmake, e.g.:
Initializing the workspace with an array of build directories, deriving the source paths from the cmakecache data and, ultimately, be closer to the machinery of cmake.

**Comment by bobbrow** [comment]
I think the extension could potentially read from the cache to get more information than it does currently, but I feel like there is a bit of a "chicken and egg" problem because first you need to tell us where the cache is and `cmake.buildDirectory` is that setting!

What I'm still unclear on is how exactly your projects tie together and what we can do to help.  We support [Variable Substitution](https://github.com/microsoft/vscode-cmake-tools/blob/develop/docs/cmake-settings.md#variable-substitution) for most settings which should enable you to set some of these settings just once (potentially at the "User" level) and have them dynamically update for each of your workspace folders, but it sounds like maybe you need a variable that we don't have yet.

**Comment by guestieng** [comment]
I don't know how cmake-tools exactly works with a multi-root workspace, but obviously it recognizes source paths when they are defined in the json "folders" array (e.g.: `"folders": [{"path": "<src. path>", ...}]`) of a workspace file *.code-workspace .
So, what I wonder is, whether the "chicken and egg" problem you formulate is really the suitable problem to start with.
Of course, I'm speaking from a perspective of not knowing the internal machinery of cmake-tools and vscode so well, but:
In the last sentence of my last comment I tried to consider it differently:
If the multi-root workspace was not set up by a source path array but a build path array instead, you would not need `cmake.buildDirectory`; ultimately, the source paths can be derived from the cmakecache, available in the build directories.
Is such an implementation possible?

More information about how the projects are exactly tied should not be relevant here from my point of view, since this can vary quite a bit depending on the workspace and workflow and it varies too much for a definite answer.
In view of this the "Variable Substitution" has not been very helpful so far, since it turns out to be too restrictive, particularly on the "User" and "Workspace" level (see also 2nd paragraph of my last comment).  

As to the question of whether another variable might be helpful, I would first ask, whether the approach described above (end of 1st paragraph) is realizable, since it would also be "closer" to cmake itself.
